### DISCUSSION TITLE
macro/2022/10/19/making-gersang-macro-and-know-how-macro-work

### DISCUSSION THREAD

**Discussion by giscus** [Code-Y-Learner/Code-Y-Learner.github.io#5] [Comments]
# macro/2022/10/19/making-gersang-macro-and-know-how-macro-work

거상 매크로 제작하고 나만의 매크로 마우스와 키보드를 만들어보자

https://code-y-learner.github.io/macro/2022/10/19/making-gersang-macro-and-know-how-macro-work.html

**Comment by nowornaver** [comment]
안녕하세요 선생님 혹시 몹을 클릭 했는지 안했는지와 저의 본진 좌표와 몬스터의 좌표는 어떻게 알아내신건지 궁금합니다.

**Reply by nowornaver** [reply]
이미지 인식은 어떤걸로 하셨나요 저는 템플릿 매칭 사용했는데 이미지 인식이 정확하게  안됍니다. 그리고 전투에 들어가면 몹을 화면에서 인식을 하는건데 전투 들어갔을때 몹의 좌표는 어떻게 구한건지도 궁금합니다..

**Reply by Code-Y-Learner** [reply]
일단  전투 때 저의 좌표와 몹의 좌표는 맵을 이용하였습니다. 맵을 검게 하면 하얀색으로 유닛이 표시되는데 단순히 픽셀을 검출하는 식으로 해도 적과 저의 본진을 구분 할 수 있습니다. 
몹을 화면에서 인식하는 거는 템플릿 매칭도 물론 좋습니다. 하지만 시연 당시 제 노트북은 10년이 넘은 노트북으로 매우 사양이 안 좋았기 때문에 단순한게 몹도 픽셀 인식으로 간단히 검출하였습니다. (픽셀 인식은 구간 내 하나의 픽셀이 이와 일치하는지 찾는 것인데 몬스터가 가지고 있는 유니크한 픽셀 하나 찾아서 넣으면 잘 작동합니다.)
그리고 몹을 클릭하는 화면과 전투화면을 서로 구별하는 것은 전투화면으로 넘어가거나 퇴각 직후 로 나누었는데 전투화면으로 넘어갈때 화면 전체가 검게 변하는 것을 검출하였습니다.

**Reply by Fish-Gogi-0525** [reply]
안녕하세요 잘 보았습니다. 혹시 해당 내용들의 공부를 공유하는 디스코드나 채널 있다면 개인적으로 받아볼 수 있나요?

**Reply by Code-Y-Learner** [reply]
관련 내용은 오래되어서 노트북 포맷하면서 다 사라졌습니다. 구축내용 자체는 크게 어렵지 않습니다. 과정은 현재 포스트와 이전에 https://code-y-learner.github.io/macro/2022/10/18/making-macro-chrome-dinosaur.html 내용이 전부입니다. 
작동 원리를 요약하자면 거상 창 모드 내에서 이미지 검출과 좌표 검색 등은 pythonautogui등을 이용해서 하고 직접적인 마우스 클릭 키보드 누름 등의 하드웨어 이벤트는 이번 포스트에서 알려드린 DDDD94687.64.dll 를 이용하는 것입니다.
혹시 pythonautogui에서 그대로 이미지 검출 후에 pythonautogui 모듈의 클릭 및 키보드 이벤트 등으로 조작하면 되지 않느냐 할 수 있는데 위에 적혀 있듯이 소프트웨어 매크로라서 관련 시도가 블락되거나 검출 될 가능성이 높습니다.

**Comment by Fish-Gogi-0525** [comment]
해당글에 답글을 이메일로 회신 주셔서 감사합니다.
답변주신 부분들 다시 체크해보도록하겠습니다

2024년 5월 8일 (수) 17:20, Code.Y.Learner ***@***.***>님이 작성:

**Comment by Code-Y-Learner** [comment]
덧붙여 제 의도를 굳이 말씀드리자면 오토 처럼 pythonautogui를 활용해서 이미지 픽셀 검출등으로 완전 자동화하는 방식을 말씀드리려 한 건 아니고 거상 다격수 메타 때문에 어쩔 수 없이 매크로 마우스 등을 쓰는데 매크로 마우스 같은거 사지말고 DDDD94687.64.dll 등을 이용해서 직접적으로 t만 누르면 1t2t3t4t 등을 원하는 간격으로 발생시키는 스크립트 등을 짜서 직접 쓰시라는 의견으로 포스트한 내용이긴 합니다.

**Comment by Fish-Gogi-0525** [comment]
그렇군요 감사합니다 매마와 해당 내용들 차이 좀 더 확인 후 진입해보도록 하겠습니다

2024년 5월 8일 (수) 17:24, Code.Y.Learner ***@***.***>님이 작성:

**Comment by gersang123** [comment]
안녕하세요.
혹시 1:1대화가능하신지요.?
ㅋㅌkym12290
꼭점 톡부탁드립니다.

**Comment by Fish-Gogi-0525** [comment]
1ㄷ1 카톡 드렸는데 답변이 없으셔서 다음에 기회가 된다면 그때 연이 되길 빌어봅니다.

2024년 5월 21일 (화) 오전 5:03, gersang123 ***@***.***>님이 작성:

**Comment by gersang-tester** [comment]
안녕하세요~
해당 부분의 질문이 있어 남깁니다.
혹시 조철레타(비올레타) 매크로탐지기는 어떻게 하셨는지?
txit010@naver.com 으로 연락방법 남겨주시면 연락드리겠습니다.

감사합니다.

**Comment by 210726** [comment]
안녕하세요 매마 쓰다가 고장이 나서 검색하다가 글을 보게 되었네요.  DDDD94687.64.dll 이용해서 하드웨어 스크립트 짜보고 있는데 무지하다보니 뭐가뭔지 잘모르겠어서 답글 남깁니다. dd소프튼가 거기서 다운 받기는 받았는데 이름도 다르고 해서 여기서 막혔습니다. 도움 받을 수 있을까요?

**Comment by HongKyoungIn** [comment]
안녕하세요. 혹시 내 본진위치 찾는거랑 몹 위치 찾는건 어떻게 하신건가요?

**Reply by HongKyoungIn** [reply]
그리고 맵 중앙으로 이동하는건 어떻게 한건가요?

**Comment by rkdlvlsfktldk** [comment]
미니맵 인식이 매우 어렵습니다 ai와 대화로 알아보려고 해도 답을 찾을 수가 없네요..

정보 살짝만 더 푸시지 ㅜㅜ

**Comment by ho7dd6-gif** [comment]
안녕하세요 싸인코드가 파이썬 사용기록과 로그들을 기록하고 보낸다고 들었습니다 아무리 파이썬으로 하드웨어 신호를 보낸다고해도 제재 당하지 않을까요?? 의견이 궁금합니다

**Comment by doi0207-bit** [comment]
왼쪽 하단 미니맵의 픽셀을 찾아서 한다고 했는데 픽셀의 rgb값을 이용한건가요..? 매 전투마다 색상이 바뀌는데 어떻게 찾아낸건지랑.. 몬스터 부대의 중앙으로가서 스킬을 쓰는데 화면이동은 어떤방식으로 한건지 궁금합니다. 어떻게 몬스터들의 중앙으로 이동해서 스킬을 쓰게 만든건지요..
